### PR TITLE
1890 - Update error encoding

### PR DESCRIPTION
**Pull request by ianmuchyri**
### What does this do?
Updates how errors are encoded in the HTTP response.
### Which issue(s) does this PR fix/relate to?
#1890 
### List any changes that modify/break current functionality
None
### Have you included tests for your changes?
No
### Did you document any new/modified functionality?
No

### CODE REVIEW

File: bootstrap/api/transport.go [hunk 1]
```diff
@@ -302,14 +301,17 @@ func encodeError(_ context.Context, err error, w http.ResponseWriter) {
 	}
 
 	if errorVal, ok := err.(errors.Error); ok {
-		w.Header().Set("Content-Type", contentType)
```
Comment (SammyOina) [inline bootstrap/api/transport.go:305]: why is this removed

File: pkg/errors/sdk_errors.go [hunk 2]
```diff
@@ -81,9 +102,18 @@ func CheckError(resp *http.Response, expectedStatusCodes ...int) SDKError {
...
+	if msg, ok := content[messageKey]; ok {
+		if msgError, ok := content[errorKey]; ok {
+			if m, ok := msg.(string); ok {
+				if e, ok := msgError.(string); ok {
+					if e != "" {
+						return NewSDKErrorWithStatus(Wrap(New(m), New(e)), resp.StatusCode)
+					}
+					return NewSDKErrorWithStatus(New(m), resp.StatusCode)
+				}
+				return NewSDKErrorWithStatus(fmt.Errorf("invalid err type: %v", msgError), resp.StatusCode)
```
Comment (SammyOina) [inline pkg/errors/sdk_errors.go]: extract to var

**Review by SammyOina — CHANGES_REQUESTED** [review]

File: pkg/errors/sdk_errors.go [hunk 3]
```diff
@@ -81,9 +102,18 @@ func CheckError(resp *http.Response, expectedStatusCodes ...int) SDKError {
 	var content map[string]interface{}
 	_ = json.Unmarshal(body, &content)
 
-	if msg, ok := content[errorKey]; ok {
-		if v, ok := msg.(string); ok {
-			return NewSDKErrorWithStatus(New(v), resp.StatusCode)
+	if msg, ok := content[messageKey]; ok {
```
Comment (dborovcanin) [inline pkg/errors/sdk_errors.go]: Since we are expecting a known format of the error, we should simplify this by unmarshalling the body to the expected struct  (rather than `map[string]interface{}`) and handling unmarshal errors, rather than using this unnecessarily complex approach.

**Review by dborovcanin — CHANGES_REQUESTED** [review]

**Review by SammyOina — APPROVED** [review]

File: bootstrap/api/transport.go [hunk 4]
```diff
@@ -303,13 +302,17 @@ func encodeError(_ context.Context, err error, w http.ResponseWriter) {
 
 	if errorVal, ok := err.(errors.Error); ok {
 		w.Header().Set("Content-Type", contentType)
-
-		errMsg := errorVal.Msg()
+		errorRes := apiutil.ErrorRes{
+			Err: "",
+			Msg: errorVal.Msg(),
+		}
```
Comment (dborovcanin) [inline bootstrap/api/transport.go]: Can we simplify error inspection by implementing JSON Marshaller in our custom error (i.e. always do just `json.NewEncoder(w).Encode(err)` with no other inspections)?

**Review by dborovcanin — CHANGES_REQUESTED** [review]

File: pkg/sdk/go/message_test.go [hunk 5]
```diff
@@ -88,15 +90,15 @@ func TestSendMessage(t *testing.T) {
...
+			err:    errors.NewSDKErrorWithStatus(unexpectedJSONEnd, http.StatusUnauthorized),
 		},
 	}
 	for desc, tc := range cases {
 		err := mfsdk.SendMessage(tc.chanID, tc.msg, tc.auth)
 		if tc.err == nil {
 			assert.Nil(t, err, fmt.Sprintf("%s: got unexpected error: %s", desc, err))
 		} else {
-			assert.Equal(t, tc.err.Error(), err.Error(), fmt.Sprintf("%s: expected error %s, got %s", desc, err, tc.err))
+			assert.Equal(t, tc.err.Error(), err.Error(), fmt.Sprintf("%s: expected error %s, got %s", desc, tc.err, err))
```
Comment (dborovcanin) [inline pkg/sdk/go/message_test.go:102]: Use `switch-default` instead of `if-else`.

File: pkg/sdk/go/tokens_test.go [hunk 6]
```diff
@@ -84,7 +84,7 @@ func TestIssueToken(t *testing.T) {
 			ok := repoCall.Parent.AssertCalled(t, "RetrieveByIdentity", mock.Anything, mock.Anything)
 			assert.True(t, ok, fmt.Sprintf("RetrieveByIdentity was not called on %s", tc.desc))
 		default:
-			assert.True(t, strings.Contains(err.Msg(), tc.err.Msg()), fmt.Sprintf("%s: expected error %s, got %s", tc.desc, tc.err, err))
+			assert.True(t, strings.Contains(err.Error(), tc.err.Msg()), fmt.Sprintf("%s: expected error %s, got %s", tc.desc, tc.err, err))
```
Comment (dborovcanin) [inline pkg/sdk/go/tokens_test.go]: You can use `assert.Contains` here to compare strings. However, we need to improve this check, since it does not include checking the status code. Probably the simplest way to do it would be to make it work with `errors.Contains`.

**Review by dborovcanin — CHANGES_REQUESTED** [review]

File: pkg/errors/errors.go [hunk 7]
```diff
@@ -49,6 +53,20 @@ func (ce *customError) Err() Error {
 	return ce.err
 }
 
+func (ce *customError) MarshalJSON() ([]byte, error) {
```
Comment (SammyOina) [inline pkg/errors/errors.go:56]: add comment on exported function

**Review by SammyOina — CHANGES_REQUESTED** [review]

File: pkg/errors/errors.go (hunk 7 above)
Comment (ianmuchyri) [reply pkg/errors/errors.go:56]: It's there: https://github.com/ianmuchyri/mainflux/blob/83b5da4b9e325d568ea6edae77fa20fd0d0194fd/pkg/errors/errors.go#L26

File: pkg/errors/errors.go [hunk 8]
```diff
@@ -21,6 +22,9 @@ type Error interface {
 
 	// Err returns wrapped error
 	Err() Error
+
+	// MarshalJSON returns a marshaled error
```
Comment (SammyOina) [inline pkg/errors/errors.go]:
```suggestion
	// MarshalJSON returns a marshaled error.
```

**Review by SammyOina — CHANGES_REQUESTED** [review]

File: Makefile [hunk 9]
```diff
@@ -17,7 +17,7 @@ USER_REPO ?= $(shell git remote get-url origin | sed -e 's/.*\/\([^/]*\)\/\([^/]
 BRANCH ?= $(shell git rev-parse --abbrev-ref HEAD  2>/dev/null || git describe --tags --abbrev=0  2>/dev/null )
 empty:=
 space:= $(empty) $(empty)
-DOCKER_PROJECT ?= $(shell echo $(subst $(space),,$(USER_REPO)_$(BRANCH)) | tr -c -s '[:alnum:][=-=]' '_' | tr '[:upper:]' '[:lower:]')
+DOCKER_PROJECT ?= $(shell echo $(subst $(space),,$(USER_REPO)_$(BRANCH)) | tr -c -s '[:alnum:][=-=]' '_')
```
Comment (dborovcanin) [inline Makefile]: This is an unrelated change and should not be here. Please be careful when rebasing.

**Review by dborovcanin — CHANGES_REQUESTED** [review]

**Review by dborovcanin — APPROVED** [review]

**Review by drasko — APPROVED** [review]
LGTM
